### ISSUE TITLE
persistent-template: Phantom self-reference is an error

### ISSUE THREAD

**Issue by lexi-lambda** [yesodweb/persistent#729]
The following code produces an error:

```haskell
import Data.Proxy (Proxy(..))

mkPersist sqlSettings [persistLowerCase|
Foo
  bar (Proxy Foo)
|]
```
```
error:
    • Exception when trying to run compile-time code:
        Foo: a self reference must be a Maybe
CallStack (from HasCallStack):
  error, called at ./Database/Persist/TH.hs:263:29 in persistent-template-2.5.3-Eo7SLuwZzwzIDXoxvwLBF1:Database.Persist.TH
      Code: template-haskell-2.11.1.0:Language.Haskell.TH.Quote.quoteExp
              persistLowerCase
              "\n\
              \Foo\n\
              \  bar (Proxy Foo)\n"
    • In the quasi-quotation:
        [persistLowerCase|
Foo
  bar (Proxy Foo)
|]
```

This isn’t really a self-reference, however, because `a` is phantom in `Proxy a`.

**Comment by parsonsmatt** [comment]
Hmmm! We should allow this. Seems like we should be able to do a [`reifyRoles`](https://www.stackage.org/haddock/lts-14.25/template-haskell-2.14.0.0/Language-Haskell-TH.html#v:reifyRoles) and if the relevant type variable is phantom then we can go straight ahead.

**Comment by parsonsmatt** [comment]
Fixed in #1256
